src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,301 @@
+use std::collections::btree_map::BTreeMap;
+use std::collections::btree_set::BTreeSet;
+use std::os::unix::fs::symlink;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::future::Future;
+use log::{debug, trace};
+
+use boxfuture::{Boxable, BoxFuture, try_future};
+use hashing::Digest;
+
+use crate::{ExecuteProcessRequest, ExecuteProcessRequestMetadata, FallibleExecuteProcessResult, MultiPlatformExecuteProcessRequest, Platform, Context};
+use crate::nailgun::nailgun_pool::NailgunProcessName;
+
+pub mod nailgun_pool;
+mod parsed_jvm_command_lines;
+
+pub use nailgun_pool::NailgunPool;
+use workunit_store::WorkUnitStore;
+use parsed_jvm_command_lines::ParsedJVMCommandLines;
+use std::fs::{read_link, remove_dir_all, remove_file};
+
+// Hardcoded constants for connecting to nailgun
+static NAILGUN_MAIN_CLASS: &str = "com.martiansoftware.nailgun.NGServer";
+static ARGS_TO_START_NAILGUN: [&str; 1] = [":0"];
+
+// We can hardcode this because we mix it into the digest in the EPR.
+// TODO(8480) This can go away when we port the fetching of the clients and servers to the rust stack.
+static NG_CLIENT_PATH: &str = "bin/ng/1.0.0/ng";
+
+///
+/// Constructs the ExecuteProcessRequest that would be used
+/// to start the nailgun servers if we needed to.
+///
+// TODO(8481) We should calculate the input_files by deeply fingerprinting the classpath.
+fn construct_nailgun_server_request(args: Vec<String>, jdk: Option<PathBuf>, platform: Platform) -> ExecuteProcessRequest {
+    let mut full_args = args;
+    full_args.push(NAILGUN_MAIN_CLASS.to_string());
+    full_args.extend(ARGS_TO_START_NAILGUN.iter().map(|a| a.to_string()));
+
+    ExecuteProcessRequest {
+        argv: full_args,
+        env: BTreeMap::new(),
+        input_files: hashing::EMPTY_DIGEST,
+        output_files: BTreeSet::new(),
+        output_directories: BTreeSet::new(),
+        timeout: Duration::new(1000, 0),
+        description: String::from("ExecuteProcessRequest to start a nailgun"),
+        unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule: hashing::EMPTY_DIGEST,
+        jdk_home: jdk,
+        target_platform: platform,
+        is_nailgunnable: true,
+    }
+}
+
+///
+/// A command runner that can run local requests under nailgun.
+///
+/// It should only be invoked with local requests.
```

Comment:
I'm not sure we can (within this file), because the ExecuteProcessRequest interface is made so that we can't tell whether we're running on local or not when we try to run. Once we have created the command runners, the rest of the code doesn't want to care about which one it's using.